src/composeWithMongooseDiscriminators.js

```diff
@@ -0,0 +1,265 @@
+import type { GraphQLFieldConfigMap } from 'graphql-compose/lib/graphql';
+import type {
+  ComposePartialFieldConfigAsObject,
+  RelationOpts,
+} from 'graphql-compose/lib/TypeComposer';
+import { Model } from 'mongoose';
+import type { TypeConverterOpts } from './composeWithMongoose';
+import { composeWithMongoose } from './composeWithMongoose';
+import { composeChildTC } from './discriminators';
+import { mergeCustomizationOptions } from './discriminators/merge-customization-options';
+import { prepareBaseResolvers } from './discriminators/prepare-resolvers/prepareBaseResolvers';
+import { reorderFields } from './discriminators/utils';
+
+const { GraphQLInterfaceType } = graphql;
+
+export type Options = {
+  test_disTypes?: boolean,
+  reorderFields?: boolean | string[], // true order: _id, DKey, DInterfaceFields, DiscriminatorFields
+  customizationOptions?: TypeConverterOpts,
+};
+
+type Discriminators = {
+  [DName: string]: any,
+};
+
+// sets the values on DKey enum TC
+function setDKeyETCValues(discriminators: Discriminators): any {
+  const values: { [propName: string]: { value: string } } = {};
+
+  for (const DName in discriminators) {
+    if (discriminators.hasOwnProperty(DName)) {
+      values[DName] = {
+        value: DName,
+      };
+    }
+  }
+
+  return values;
+}
+
+// creates an enum from discriminator names
+// then sets this enum type as the discriminator key field type
+function createAndSetDKeyETC(dTC: DiscriminatorTypeComposer, discriminators: Discriminators) {
+  const DKeyETC = EnumTypeComposer.create({
+    name: `EnumDKey${dTC.getDBaseName()}${dTC.getDKey()[0].toUpperCase() +
+      dTC.getDKey().substr(1)}`,
+    values: setDKeyETCValues(discriminators),
+  });
+
+  // set on Output
+  dTC.extendField(dTC.getDKey(), {
+    type: () => DKeyETC,
+  });
+
+  // set on Input
+  dTC.getInputTypeComposer().extendField(dTC.getDKey(), {
+    type: () => DKeyETC,
+  });
+
+  return DKeyETC;
+}
+
+function getBaseTCFieldsWithTypes(baseTC: TypeComposer) {
+  const baseFields = baseTC.getFieldNames();
+  const baseFieldsWithTypes: GraphQLFieldConfigMap<any, any> = {};
+
+  for (const field of baseFields) {
+    baseFieldsWithTypes[field] = baseTC.getFieldConfig(field);
+  }
+
+  return baseFieldsWithTypes;
+}
+
+function createDInterface(baseModelTC: DiscriminatorTypeComposer): GraphQLInterfaceType {
+  return new GraphQLInterfaceType({
+    name: baseModelTC.getDBaseName(),
+
+    resolveType: (value: any) => {
+      const childDName = value[baseModelTC.getDKey()];
+      const sComposer = baseModelTC.getGQC(); // get schemaComposer
+
+      if (childDName) {
+        return sComposer.getTC(childDName).getType();
+      }
+
+      // as fallback return BaseModelTC
+      return sComposer.getTC(baseModelTC.getTypeName()).getType();
+    },
+    // hoisting issue solved, get at time :)
+    fields: () => getBaseTCFieldsWithTypes(baseModelTC),
+  });
+}
+
+export class DiscriminatorTypeComposer extends TypeComposer {
```

Comment:
Please move `DiscriminatorTypeComposer` class to its own file src/discriminators/DiscriminatorTypeComposer.js